File: src/main/java/at/hrechny/predictionsbot/util/HashUtils.java
package at.hrechny.predictionsbot.util;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import lombok.SneakyThrows;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

@Service
public class HashUtils {

  @Value("${secrets.telegramKey}")
  private String telegramKey;

  @SneakyThrows
  public String getHash(String originalString) {
    final MessageDigest digest = MessageDigest.getInstance("SHA-256");
    digest.update(telegramKey.getBytes(StandardCharsets.UTF_8));
    final byte[] hashbytes = digest.digest(originalString.getBytes(StandardCharsets.UTF_8));
    return bytesToHex(hashbytes);
  }

  private static String bytesToHex(byte[] hash) {
    StringBuilder hexString = new StringBuilder(2 * hash.length);
    for (byte b : hash) {
      String hex = Integer.toHexString(0xff & b);
      if (hex.length() == 1) {
        hexString.append('0');
      }
      hexString.append(hex);
    }
    return hexString.toString();
  }

}

File: src/main/kotlin/at/hrechny/predictionsbot/util/HashUtils.kt
package at.hrechny.predictionsbot.util

import io.micronaut.context.annotation.Value
import jakarta.inject.Singleton
import java.nio.charset.StandardCharsets
import java.security.MessageDigest

@Singleton
class HashUtils(
    @param:Value("\${secrets.telegramKey}")
    private val telegramKey: String,
) {
    fun getHash(originalString: String): String {
        val digest = MessageDigest.getInstance("SHA-256")
        digest.update(telegramKey.toByteArray(StandardCharsets.UTF_8))
        val hashBytes = digest.digest(originalString.toByteArray(StandardCharsets.UTF_8))
        return bytesToHex(hashBytes)
    }

    private companion object {
        fun bytesToHex(hash: ByteArray): String {
            val hexString = StringBuilder(2 * hash.size)
            for (byte in hash) {
                val hex = Integer.toHexString(0xff and byte.toInt())
                if (hex.length == 1) {
                    hexString.append('0')
                }
                hexString.append(hex)
            }
            return hexString.toString()
        }
    }
}
